### PR TITLE
feat: integrate ada handle

### PR DESCRIPTION
**Pull request by sk-tenba**
https://app.asana.com/0/1203842063837585/1203857184580238

### CODE REVIEW

**Comment by sk-tenba** [comment]
Since this could not be tested on preview, I could only show the UI after using some mock handle

**Comment by sk-tenba** [comment]
Screenshot from 2023-02-28 17-32-57

**Comment by sk-tenba** [comment]
Screenshot from 2023-02-28 17-27-39

**Comment by sk-tenba** [comment]
Screenshot from 2023-03-02 14-59-38

**Comment by sk-tenba** [comment]
Screenshot from 2023-03-02 15-57-24

**Comment by sk-tenba** [comment]
Screenshot from 2023-03-02 18-49-44

**Comment by sk-tenba** [comment]
Screenshot from 2023-03-02 18-49-35

**Comment by sk-tenba** [comment]
Screenshot from 2023-03-02 18-49-27

**Comment by sk-tenba** [comment]
[Screencast from 03-03-2023 10:40:36.webm](https://user-images.githubusercontent.com/110231356/222625915-509ad6d3-28e8-4e64-89f4-2a4e6158aaa2.webm)

File: containers/PageHome/containers/NavBar/containers/ButtonWalletNavbar/containers/ButtonWalletOptions/index.tsx [hunk 1]
```diff
@@ -47,13 +50,23 @@ export default function ButtonWalletOptions({
...
+                  />
+                ) : (
+                  <Typography.Div
+                    size="heading6"
+                    color="ink"
+                    className={styles.handle}
+                    content={`$${data}`}
+                    maxLines={1}
+                  />
+                )}
```
Comment (sk-kitsune) [inline containers/PageHome/containers/NavBar/containers/ButtonWalletNavbar/containers/ButtonWalletOptions/index.tsx:68]: Change to `Typography.Span`.

The children inside a `Typography.Div` should usually be `<span>`, `Typography.Span`, or anything else with `display: inline` (like `InlineAddress`).

Comment (sk-kitsune) [reply containers/PageHome/containers/NavBar/containers/ButtonWalletNavbar/containers/ButtonWalletOptions/index.tsx:68]: move `maxLines={1}` to the parent component

File: containers/PageProfile/containers/PanelActivities/components/ActivityViewer/components/ActionBack.tsx [hunk 2]
```diff
@@ -15,11 +16,16 @@ type Props = {
...
   return (
     <Flex.Col className={className} style={style} gap="8px">
       <Typography.Div size="bodySmall" color="ink80">
         <Typography.Span fontWeight="semibold" color="ink">
-          <InlineAddress value={value.createdBy} length="short" />
+          {error != null || data == undefined ? (
+            <InlineAddress value={value.createdBy} length="short" />
+          ) : (
+            `$${data}`
+          )}
```
Comment (sk-kitsune) [inline containers/PageProfile/containers/PanelActivities/components/ActivityViewer/components/ActionBack.tsx:28]: Actually, `InlineAddress` should handle the logic of ADA handler.

Avoid shotgun surgery.

---

I recommend:

`<InlineAddress value={value.createdBy} length="short" allowAdaHandle={true} />`

Comment (sk-kitsune) [reply containers/PageProfile/containers/PanelActivities/components/ActivityViewer/components/ActionBack.tsx:28]: https://en.wikipedia.org/wiki/Shotgun_surgery

Comment (sk-kitsune) [reply containers/PageProfile/containers/PanelActivities/components/ActivityViewer/components/ActionBack.tsx:28]: change

```
type Props = {
  className?: string;
  style?: React.CSSProperties;
  value: Address;
};
```

to 


```
type Props = {
  className?: string;
  style?: React.CSSProperties;
  value: Address;
  enableAdaHandle?: boolean; // default: true
};
```

Comment (sk-kitsune) [reply containers/PageProfile/containers/PanelActivities/components/ActivityViewer/components/ActionBack.tsx:28]: maybe a required param is a safer choice, i.e. `enableAdaHandle: boolean;`

File: modules/common-hooks/hooks/useAdaHandle.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    quantity: string;
+  }[];
+  stake_address: string;
+  type: string;
+  script: boolean;
+};
+
+export default function useAdaHandle(address?: string) {
+  const policyID = NEXT_PUBLIC_HANDLE_POLICY_ID;
+  const { data, error } = useSWR(address, async (address) => {
```
Comment (sk-kitsune) [inline modules/common-hooks/hooks/useAdaHandle.ts:22]: The first argument of the useSWR, i.e. the resource key, must be globally unique.

Attach a UUID to make it unique.

---

🔴 **Blocking issue**: this PR cannot be merged until this issue is resolved.

Comment (sk-kitsune) [reply modules/common-hooks/hooks/useAdaHandle.ts:22]: If another hook also uses the same key (i.e. `address`) then one hook will see the response of another hook.

File: modules/common-hooks/hooks/useAdaHandle.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    if (address === "" || address == null) {
+      return undefined;
+    }
```
Comment (sk-kitsune) [inline modules/common-hooks/hooks/useAdaHandle.ts:25]: `if (!address) return undefined`

File: modules/common-hooks/hooks/useAdaHandle.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    const response = await fetch(
+      `${NEXT_PUBLIC_BLOCKFROST_URL}/addresses/${address}`,
+      {
+        headers: {
+          project_id: NEXT_PUBLIC_BLOCKFROST_PROJECT_ID,
+          "Content-Type": "application/json",
+        },
+      }
+    );
```
Comment (sk-kitsune) [inline modules/common-hooks/hooks/useAdaHandle.ts:34]: https://www.notion.so/shinka-network/RFC-TD-Guidelines-on-using-Server-Side-Props-SSP-48551043ee154ae0bac21ea0625d0cf3?pvs=4#2c0c570e30b841e886ac68f36ba896c8

As a convention, please define `httpGetAdaHandle`.

File: modules/common-hooks/hooks/useAdaHandle.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    const data: Response = await response.json();
+
+    const handles = data.amount
+      .filter(({ unit }) => unit.includes(policyID))
+      .map(({ unit }) => {
+        const hexName = unit.replace(policyID, "");
+        const utf8Name = Buffer.from(hexName, "hex").toString("utf8");
```
Comment (sk-kitsune) [inline modules/common-hooks/hooks/useAdaHandle.ts:41]: where is `Buffer` imported from? are you sure it is available on browser?

File: containers/PageProfile/containers/PanelActivities/components/ActivityViewer/components/ActionBack.tsx (hunk 2 above)
Comment (sk-kitsune) [reply containers/PageProfile/containers/PanelActivities/components/ActivityViewer/components/ActionBack.tsx:28]: 🔴 **Blocking issue**: this PR cannot be merged until this issue is resolved.

**Review by sk-kitsune — CHANGES_REQUESTED** [review]
⛔ **Changes requested**: there are blocking issues.

File: config/client.ts [hunk 7]
```diff
@@ -132,3 +132,10 @@ assert(
   "NEXT_PUBLIC_IPFS_GATEWAY_ORIGIN must starts with " +
     "'http://' or 'https://' and must not end with '/'"
 );
+
+export const NEXT_PUBLIC_HANDLE_POLICY_ID: string =
```
Comment (sk-shishi) [inline config/client.ts:136]: Rename this to `NEXT_PUBLIC_ADA_HANDLE_POLICY_ID`

**Comment by sk-shishi** [comment]
Please don't merge this. This client-side solution definitely won't scale, as it will send a request to Blockfrost for every address shown on our front-end. It can easily reach the Blockfrost rate limit (per IP).

**Comment by sk-shishi** [comment]
@sk-kitsune @sk-tenba

I created a new task for the indexing part https://app.asana.com/0/1203842063837585/1204114484985397/f.

To be frank, this PR can be merged (so we can do some UI testing) while waiting for the indexing task to be finished. However, to reduce the number of Blockfrost calls, we need to at least disable all `useSWR` auto-revalidation for that Address->Handle lookup call.

Fyi, Blockfrost rate limit per IP

> We limit an end user, distinguished by IP address, to 10 requests per second. On top of that, we allow each user to send burst of 500 requests, which cools off at rate of 10 requests per second.

**Comment by sk-shishi** [comment]
@sk-tenba With the upcoming API, we can get the ADA handles of multiple addresses in a single API call. Currently, we're sending one request for each address, which is inefficient. Do you want to handle it now (in this PR) or we'll do it later?

**Comment by sk-tenba** [comment]
@sk-shishi I think sending multiple addresses would introduce complicated code refactoring. Let's do it later.

**Pull request closed without merging**